### ISSUE TITLE
Default cache use wrong cached values

### ISSUE THREAD

**Issue by aljim** [jayway/JsonPath#234]
```
Map<String, String> context = Maps.newHashMap();
context.put("key", "first");
Object value = JsonPath.read(context, "concat(\"/\", $.key)");
assertThat(value).isEqualTo("/first");
context.put("key", "second");
value = JsonPath.read(context, "concat(\"/\", $.key)");
assertThat(value).isEqualTo("/second");
```

That test fails, the cache seems to do something very odd with concat, if not using concat it does not fail, but using concat the cache uses the same value, even that the map where is reading changed.

**Comment by jochenberger** [comment]
I can confirm the issue.
`FunctionPathToken` and `Parameter` cache the result from the first evaluation and return it also if the input has changed. @mgreenwood1001, would you like to have a look?

**Comment by mgreenwood1001** [comment]
I'll take a look this evening and provide a fix.

**Comment by jochenberger** [comment]
Will be fixed in 2.3.0

**Comment by mborgraeve** [comment]
Hey there ! I'm using 2.4.0 and i'm still having this issue, is that expected ?
